core/src/main/java/com/backblaze/b2/json/B2JsonInitializedTypeHandler.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2019, Backblaze Inc. All Rights Reserved.
+ * License https://www.backblaze.com/using_b2_code.html
+ */
+
+package com.backblaze.b2.json;
+
+import com.backblaze.b2.util.B2Preconditions;
+
+/**
+ * Base class for all implementations of B2JsonTypeHandler that have an initialize() method.
+ *
+ * (De)serialization is expected to be fast, so each implementation class gathers
+ * information it needs when it's set up, so when it's time to run it has all of
+ * the necessary information at hand.  This gets tricky because dependencies between
+ * handlers may have loops.
+ *
+ * The plan is to initialize in two phases:
+ *
+ * First, the constructor, which must not depend on any other handlers, and which
+ * must gather all of the information that other handlers will need.
+ *
+ * Second, the initialize() method does any work that needs information from other
```

Comment:
Since some code will be calling getUninitializedHandler() and warned to not count on things that aren't initialized yet, how can we communicate what that will be for type handlers the caller doesn't know about?  Is it more about which calls can't be called before initialize has been called?  And, does that leak into the B2JsonTypeHandler interface since serialize & deserialize are probably things that shouldn't be called until the class is initialized?

Reply:
I'm wondering if we should just make initialize() a method on the base interface.

Reply:
I'd be OK with moving `initialize()` up.  Would that make things clearer for you?

The rule is that constructors must assume that no handler they see is initialized.

Reply:
Oh.  Now I remember.  Moving the `initialize()` method up would change the interface, and break anybody who had their own implementation of it.  So I rejected that.

Reply:
(brian and i talked separately.)

i was concerned that we were changing the meaning of the handler map in constructors and that these changes only give two-phase initialization to handlers that inherit from a particular abstract class.  i missed that we already don't pass the handler map to custom handlers.

so i'm ok leaving initialize as part of this subclass.